### ISSUE TITLE
some questions for the Intent-guided process and contrastive learning

### ISSUE THREAD

**Issue by Kris20xx** [qyp9909/InDiRec#3]
Thank you for your excellent work on InDiRec! I am currently reproducing your paper and noticed 1 potential issue in the `get_sequences_in_same_cluster` function, along with unexpected performance changes after I modified it. I would greatly appreciate your insights on this.  

### Problem Description  1
In the original implementation of `get_sequences_in_same_cluster(input_sequence)`:  
```python
# Original code snippet (with core logic retained)
cluster_idx = int(I[0][0])  
sequence_ids_in_cluster = cluster.cluster2sequences[cluster_idx]
sequences_in_cluster = [self.sequences_list[seq_id] for seq_id in sequence_ids_in_cluster]
random_sequences = random.choices(sequences_in_cluster, k=batch_size)
return random_sequences
```  
It seems that this logic uses the cluster index of **only the first sample in the batch** (`I[0][0]`) to sample sequences for the entire batch. This would mean all samples in a batch share the same intent-guided vectors (from the same cluster), which I believe is inconsistent with the paper’s design—each sequence in the batch should independently find its own corresponding cluster and sample intent-guided sequences from its own cluster.  

### My Modification  1
To fix this, I revised the function to sample sequences per sample in the batch:  
```python
# Modified code
random_sequences = []
for i in range(batch_size):
    cluster_idx = int(I[i][0])
    seq_ids = cluster.cluster2sequences[cluster_idx]
    sampled_seq = random.choice(seq_ids)
    random_sequences.append(self.sequences_list[sampled_seq])
return random_sequences
```  

### Unexpected Result 1 
After this modification, the evaluation metrics **decreased slightly** instead of improving:  

| Metric       | Original Result | Modified Result |
|--------------|-----------------|-----------------|
| HR@5         | 0.0679          | 0.0675          |
| NDCG@5       | 0.0488          | 0.0475          |
| HR@10        | 0.0943          | 0.0929          |
| NDCG@10      | 0.0574          | 0.0556          |
| HR@20        | 0.1286          | 0.1260          |
| NDCG@20      | 0.0660          | 0.0640          |  

### Questions for You  1
1. Did I misunderstand the design logic of `get_sequences_in_same_cluster`? Is using the first sample’s cluster for the entire batch an intentional design choice (e.g., for efficiency or regularization)?  
2. Why might the per-sample cluster sampling lead to slightly lower metrics? Could it be related to batch-level regularization, noise in cluster sampling, or other hyperparameters I missed?  

### Questions for You  2
I have another brief question regarding the contrastive loss implementation.

In Section 3.3 of your paper, it states that the output from the diffusion reverse process (the augmented view embedding $\hat{e}_0$) should first be passed through the sequence encoder (Eq. 8) to obtain the representation $h_2$, which is then used in the contrastive loss.

However, in the code, 
```python
x0_hat_out = self.model.sample_from_reverse_process(s_hat[:,-1,:]) 
```
it appears that the diffusion reverse output x0_hat_out is directly used as $h_2$ in the contrastive loss calculation, without going through the sequence encoder.Is this an intentional simplification in the code, or did I misinterpret the data flow?

**Comment by SHIREN234** [comment]
thank u for your finding    this saves a lot of time for me!!!! Thank u

**Comment by qyp9909** [comment]
@Kris20xx Thank you for the detailed reproduction and for pointing these out, and sorry for the delayed response.

For the first question, your interpretation is correct. The intended behavior is that each sequence in a mini-batch independently queries its corresponding intent cluster and samples a guidance sequence from that cluster. Using `I[0][0]` for the entire batch is an implementation oversight in the released code rather than an intentional batch-level regularization or efficiency design. Your per-sample modification is consistent with the intended intent-guided sampling procedure.

Regarding the slight performance decrease after this modification, I would not interpret a single run as evidence that the batch-level sampling is preferable. The correction changes the guidance distribution seen during training, while K-means assignment, within-cluster sampling, diffusion noise, and model initialization all introduce stochasticity. In addition, cluster assignments are not necessarily perfect, so stronger per-sample conditioning does not necessarily lead to a monotonic improvement in every run. We will correct the released implementation to use per-sample cluster sampling and evaluate the corrected version across multiple random seeds to determine whether this small difference is statistically meaningful.

For the second question, you also identified a genuine discrepancy between the description in Section 3.3 and the released implementation.

In the released code, the diffusion target is obtained by

`x0_gt = calculate_x0(target)`

where `calculate_x0()` applies the item embedding layer to the target item. Therefore, the diffusion process operates on a `B × D` target-item embedding rather than a `B × L × D` sequence embedding. The reverse process consequently outputs another `B × D` vector, `x0_hat_out`, which is directly used in the contrastive objective together with the final `B × D` representation of the original sequence produced by the sequence encoder.

This also explains why `x0_hat_out` is not passed through the sequence encoder again in the released implementation. The current sequence encoder takes a sequence of item IDs as input and internally applies item and positional embeddings, whereas `x0_hat_out` is already a continuous `B × D` vector rather than an item-ID sequence. An additional encoder pass as described in Section 3.3 is therefore not part of the current implementation.

Thus, your reading of the code is correct: the additional sequence-encoder step described for the generated view in Section 3.3 is not present in the released implementation. We appreciate you pointing out this inconsistency.

Thank you again for carefully examining the implementation.
